xpublish_edr/geometry/common.py

```diff
@@ -19,3 +28,106 @@ def is_regular_xy_coords(ds: xr.Dataset) -> bool:
     Check if the dataset has 2D coordinates
     """
     return coord_is_regular(ds.cf["X"]) and coord_is_regular(ds.cf["Y"])
+
+
+def dataset_crs(ds: xr.Dataset) -> pyproj.CRS:
+    grid_mapping_names = ds.cf.grid_mapping_names
+    if len(grid_mapping_names) == 0:
+        # Default to WGS84
+        return pyproj.crs.CRS.from_epsg(4326)
+    if len(grid_mapping_names) > 1:
+        raise ValueError(f"Multiple grid mappings found: {grid_mapping_names!r}!")
+    (grid_mapping_var,) = tuple(itertools.chain(*ds.cf.grid_mapping_names.values()))
+
+    grid_mapping = ds[grid_mapping_var]
+    return pyproj.CRS.from_cf(grid_mapping.attrs)
+
+
+def project_geometry(ds: xr.Dataset, geometry_crs: str, geometry: Geometry) -> Geometry:
+    """
+    Get the projection from the dataset
+    """
+    data_crs = dataset_crs(ds)
+
+    transformer = transformer_from_crs(
+        crs_from=geometry_crs,
+        crs_to=data_crs,
+        always_xy=True,
+    )
+    return transform(transformer.transform, geometry)
+
+
+def project_dataset(ds: xr.Dataset, query_crs: str) -> xr.Dataset:
+    """
+    Project the dataset to the given CRS
+    """
+    data_crs = dataset_crs(ds)
+    target_crs = pyproj.CRS.from_string(query_crs)
+    if data_crs == target_crs:
+        return ds
+
+    transformer = transformer_from_crs(
+        crs_from=data_crs,
+        crs_to=target_crs,
+        always_xy=True,
+    )
+
+    # TODO: Handle rotated pole
+    cf_coords = target_crs.coordinate_system.to_cf()
+
+    # Get the new X and Y coordinates
+    target_y_coord = next(coord for coord in cf_coords if coord["axis"] == "Y")
+    target_x_coord = next(coord for coord in cf_coords if coord["axis"] == "X")
+
+    X = ds.cf["X"]
+    Y = ds.cf["Y"]
+
+    # Transform the coordinates
+    # If the data is vectorized, we just transform the points in full
+    # TODO: Handle 2D coordinates
+    if len(X.dims) > 1 or len(Y.dims) > 1:
+        raise NotImplementedError("Only 1D coordinates are supported")
+
+    x_dim = X.dims[0]
+    y_dim = Y.dims[0]
+    if x_dim == VECTORIZED_DIM and y_dim == VECTORIZED_DIM:
+        x = X
+        y = Y
+        target_dims: tuple = (VECTORIZED_DIM,)
+    else:
+        # Otherwise we need to transform the full grid
+        # TODO: Is there a better way to handle this? this is quite hacky
+        var = [d for d in ds.data_vars if x_dim and y_dim in ds[d].dims][0]
+        var_dims = ds[var].dims
+        if var_dims.index(x_dim) < var_dims.index(y_dim):
+            x, y = xr.broadcast(X, Y)
+            target_dims = (x_dim, y_dim)
+        else:
+            x, y = xr.broadcast(Y, X)
+            target_dims = (y_dim, x_dim)
```

Comment:
```suggestion
    x, y = xr.broadcast(x, y)
    target_dims = x.dims
```

just always broadcast. If `x.dims == y.dims` (the vectorized case) this does nothing.

And this looks wrong?
<img width="238" alt="image" src="https://github.com/user-attachments/assets/49abf785-61cb-44ad-8e64-7cd61d72ac31">

Reply:
Does the order matter? Like if the dims in the dataset are `(y, x)` do we need to make sure to broadcast (`broadcast(Y, X)`) in that order too?

Reply:
it has some heuristic but does it matter?

Reply:
Perhaps we should add an explicit transpose so that we can guarantee a returned dimension order